### PR TITLE
FC3 migration to Module

### PR DESCRIPTION
**Pull request by jdahlblom**

### CODE REVIEW

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+FC3:define8BitFloatFromGetter("FC3_SIDE_DEVIATION", function()
+	return side_deviation or 0
+end, { -1, 1 }, "Float", "Side Deviation")
+
+FC3:define8BitFloatFromGetter("FC3_SLIP_BALL_POSITION", function()
+	return slip_ball_position or 0
+end, { -1, 1 }, "Float", "Slip Ball Position")
+
+--Externals
+FC3:defineIntegerFromGetter("EXT_SPEED_BRAKE_RIGHT", function()
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:438]: You can use defineFloatFromDrawArgument for these calls

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+	return math.floor(LoGetAircraftDrawArgumentValue(182) * 65535)
+end, 65535, "External Aircraft Model", "Right Speed Brake")
+
+FC3:defineIntegerFromGetter("EXT_SPEED_BRAKE_LEFT", function()
+	return math.floor(LoGetAircraftDrawArgumentValue(184) * 65535)
+end, 65535, "External Aircraft Model", "Left Speed Brake")
+
+FC3:defineIntegerFromGetter("EXT_POSITION_LIGHT_LEFT", function()
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:446]: Remember to use defineBitFromDrawArgument for these

File: Scripts/DCS-BIOS/lib/modules/Module.lua [hunk 3]
```diff
@@ -98,6 +98,29 @@ function Module:defineFloat(identifier, arg_number, limits, category, descriptio
...
+	-- same as defineFloat, but only allocates an 8-bit int
+	local max_value = 255
+	local intervalLength = limits[2] - limits[1]
+	local alloc = self:allocateInt(max_value)
+
+	self:addExportHook(function()
+		alloc:setValue(((func() - limits[1]) / intervalLength) * 255)
+	end)
+
+	local alloc = moduleBeingDefined.memoryMap:allocateInt({ maxValue = 255 })
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/Module.lua:111]: This and the next few lines look like old code that should be deleted?

File: Scripts/DCS-BIOS/lib/modules/Module.lua [hunk 4]
```diff
@@ -98,6 +98,29 @@ function Module:defineFloat(identifier, arg_number, limits, category, descriptio
 	return control
 end
 
+function Module:define8BitFloatFromGetter(identifier, func, limits, category, description)
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/Module.lua:101]: Remember to add documentation, and also tests for any new control

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+		return true
+	elseif plane == "Su-25T" then
+		return true
+	else
+		return false
+	end
+end
+
+FC3:addExportHook(function()
+	local selfdata = LoGetSelfData()
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:145]: I would break all these into separate export hooks personally, since none of them rely on each other

**Review by charliefoxtwo — CHANGES_REQUESTED** [review]
We should really get the `define8BitFloatFromGetter` added to the module tests before merging this so we can make sure it's tested and correct.

**Comment by jdahlblom** [comment]
> We should really get the `define8BitFloatFromGetter` added to the module tests before merging this so we can make sure it's tested and correct.

OK. So that is quite a lot of work isn't it?

File: Scripts/DCS-BIOS/lib/modules/Module.lua (hunk 4 above)
Comment (jdahlblom) [reply Scripts/DCS-BIOS/lib/modules/Module.lua:101]: fixed

**Comment by charliefoxtwo** [comment]
It should be really similar to the tests for defineIntegerFromGetter. Check out `IntegerFromGetterTests.lua`

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 6]
```diff
@@ -435,60 +435,13 @@ FC3:define8BitFloatFromGetter("FC3_SLIP_BALL_POSITION", function()
...
-
-FC3:defineIntegerFromGetter("EXT_WOW_LEFT", function()
-	if LoGetAircraftDrawArgumentValue(6) > 0 then
-		return 1
-	else
-		return 0
-	end
-end, 1, "External Aircraft Model", "Weight ON Wheels Left Gear")
+FC3:defineBitFromDrawArgument("EXT_SPEED_BRAKE_RIGHT", 182, "External Aircraft Model", "Right Speed Brake")
+FC3:defineBitFromDrawArgument("EXT_SPEED_BRAKE_LEFT", 184, "External Aircraft Model", "Left Speed Brake")
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:439]: these two should be `defineFloatFromDrawArgument`. The max value of `defineBitFromDrawArgument` is 1, and these should output 0-65535

File: Scripts/DCS-BIOS/test/AircraftTest.lua [hunk 7]
```diff
@@ -80,7 +84,7 @@ function TestAircraft:validateControlNames(module_name, documentation)
 			--   start with a letter
 			--   end with a letter or number
 			--   contain only letters, numbers, and underscores
-			local identifier_pattern = "^%u[%u%d_]*[%u%d]$"
+			local identifier_pattern = "^[%u%d_]*[%u%d]$"
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/test/AircraftTest.lua:87]: I'd make this change in a separate PR so we can fix the build right away and it doesn't depend on this PR, which will require a bit more work

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua (hunk 5 above)
Comment (jdahlblom) [reply Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:145]:
```
	local plane = GetPlaneName()

	local alt_sea = LoGetAltitudeAboveSeaLevel() or 0
	local alt_ground = LoGetAltitudeAboveGroundLevel() or 0
	local alt = alt_sea or 0
	local vvi = LoGetVerticalVelocity() or 0
	local ias = LoGetIndicatedAirSpeed() or 0
	local tas = LoGetTrueAirSpeed() or 0
	local mach = LoGetMachNumber() or 0
	local fuel_all = LoGetFuelAll() or 0
	local aoa = LoGetAngleOfAttack() or 0
	glide_deviation = LoGetGlideDeviation()
	side_deviation = LoGetSideDeviation()
	slip_ball_position = LoGetSlipBallPosition()
``` 
All these into their own hooks?

**Comment by jdahlblom** [comment]
I see these changes in FC3 json because of the new define8BitFloatFromGetter().  I flew master and this and the values all seems to work. Compared them when on runway also.

fc3

File: Scripts/DCS-BIOS/lib/modules/Module.lua (hunk 4 above)
Comment (charliefoxtwo) [reply Scripts/DCS-BIOS/lib/modules/Module.lua:108]: Will you be adding tests to ModuleTests?

Comment (jdahlblom) [reply Scripts/DCS-BIOS/lib/modules/Module.lua:108]: Yes I will try.

**Comment by charliefoxtwo** [comment]
> I see these changes in FC3 json because of the new define8BitFloatFromGetter(). I flew master and this and the values all seems to work. Compared them when on runway also.

Good catch. I wasn't aware of the value_range field in the json. We should look at what that's actually for. If that's the dcs output, then it doesn't really serve any value and we can remove it.

**Comment by jdahlblom** [comment]
> We should really get the `define8BitFloatFromGetter` added to the module tests before merging this so we can make sure it's tested and correct.

Done!

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+		maxFuel = 9400
+	end
+
+	local barf = math.ceil((fuel / maxFuel) * 16)
+	if barf > 16 then
+		barf = 16
+	end
+	if barf == nil then
+		return
+	end
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua]: I don't believe this is needed. Looks like a lot of the functions have these unnecessary nil checks; we should remove them

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+end
+
+local function allowMach(plane)
+	if plane == "A-10A" then
+		return false
+	elseif plane == "Su-25T" then
+		return false
+	else
+		return true
+	end
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:129]: this can be condensed to
```suggestion
	 return plane ~= "A-10A" and plane ~= "Su-25T
```

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+end
+
+local function allowTAS(plane)
+	if plane == "Su-25" then
+		return true
+	elseif plane == "Su-25T" then
+		return true
+	else
+		return false
+	end
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/FC3.lua:139]: this can be condensed to
```suggestion
	return plane == "Su-25" or plane == "Su-25T"
```

**Review by charliefoxtwo — DISMISSED** [review]
A couple minor opportunities to clean things up a bit more, but otherwise looks good!

**Review by charliefoxtwo — APPROVED** [review]
